### PR TITLE
Truchet

### PR DESCRIPTION
**Pull request by johanvts**
Thanks for making this and for a fantastic paper.

I wanted to see a larger patch of the Truchet style pattern overlay suggested by Daniel Piker here: https://mathstodon.xyz/@Danpiker/110062396666001681
So I added his Truchet style pattern to the tool. 

I also made a few more tweaks:
- Added some documentation to the code
- Control of the stroke color and on/off stroke toggle (a little wonky as it must be re-toggled with color changes for now)
- Control of background color.

I'm hosting the modified tool here: https://www.johansivertsen.com/customhatviz/app.html
Some example outputs showing the interlocking fylfots and the Truchet pattern together are here: https://www.johansivertsen.com/post/monotile/

Feel free to pick anything of use and discard the rest.

### CODE REVIEW

**Comment by isohedral** [comment]
I like the ideas you offer, but I feel they're not quite ready for incorporation. In particular:
* I don't like that the tile markings are on top of everything else. Ideally, they should be sandwiched above the hat fills and below the hat outlines. Definitely below the metatile outlines.
* The markings need to be included in the SVG output.
* There are a few other minor things (e.g., changing whitespace everywhere).

Basically I think that the underlying architecture of my code needs a bit of cleaning up before a feature like this can be added cleanly. I'm going to ponder a suite of changes along those lines, at which point we can revisit your features.

**Comment by isohedral** [comment]
I've checked in revised code that supports tile markings more cleanly, if you want to take another look. Apart from additions to the UI, rendering markings should affect the `draw` and `buildSVGDefs` methods in the `HatTile` class.

**Comment by johanvts** [comment]
Thank you. Yes I admit I took the shortest route to just getting the visual results I wanted. Your revised architecture looks much improved. I think the function below is really the only part not superseded by your new changes. I will just leave it here as a easy starting point for my future self (or anyone else). Thanks again.

````
function drawTruchet(shape,T,f,s,w)
{
    if( f != null ) { // Set fill color or disable fill
	fill( ...f );
    } else {
	noFill();
    }
    if( s != null ) { // Set stroke color, stroke weight or disable stroke
	stroke( ...s );
	strokeWeight( w * lw_scale );
    } else {
	noStroke();
    }
    //Bottom disk
    //The center of the bottom disk
    let c = hexPt(-2,4);
    //Left arm
    let l = pt(lerp(shape[12].x,shape[0].x,0.5),lerp(shape[12].y,shape[0].y,0.5));		      
    //Right arm
    let r = pt(lerp(shape[9].x,shape[10].x,0.5),lerp(shape[9].y,shape[10].y,0.5));

    beginShape();
    for(let p of [r,shape[10],shape[11],shape[12],l]){
	const pt = transPt(T,p);
	vertex(pt.x,pt.y);
    }
    for(let arcPoint of arcPoints(c.x,c.y,r.x,r.y,l.x,l.y,20))
    {
	const pt = transPt(T,arcPoint);
	vertex(pt.x, pt.y);
    }
    endShape();    

    // Belt
    beginShape();
    let startpoint = pt(lerp(shape[0].x,shape[1].x,0.5),lerp(shape[0].y,shape[1].y,0.5));
    for(let p of [startpoint,shape[1],shape[2],shape[3],shape[4]]){
	const tp = transPt(T,p);
	vertex(tp.x,tp.y);
    }
    //Upper arc
    c = shape[5]
    l = pt(lerp(shape[5].x,shape[4].x,0.5),lerp(shape[5].y,shape[4].y,0.5));
    r = pt(lerp(shape[5].x,shape[6].x,0.5),lerp(shape[5].y,shape[6].y,0.5));
    for(let arcPoint of arcPoints(c.x,c.y,l.x,l.y,r.x,r.y,20))
    {
	const pt = transPt(T,arcPoint);
	vertex(pt.x, pt.y);
    }
    for(let p of [shape[6],shape[7],shape[8]]){
	const tp = transPt(T,p);
	vertex(tp.x,tp.y);
    }

    //lower main arc
    l = pt(lerp(shape[4].x,shape[0].x,0.5),lerp(shape[4].y,shape[0].y,0.5));		      
    r = pt(lerp(shape[8].x,shape[9].x,0.5),lerp(shape[8].y,shape[9].y,0.5));

    for(let arcPoint of arcPoints(c.x,c.y,r.x,r.y,l.x,l.y,20))
    {
	const pt = transPt(T,arcPoint);
	vertex(pt.x, pt.y);
    }
    //lower minor arc
    c = shape[0];
    r = l;
    l = startpoint;

    for(let arcPoint of arcPoints(c.x,c.y,r.x,r.y,l.x,l.y,20))
    {
	const pt = transPt(T,arcPoint);
	vertex(pt.x, pt.y);
    }
    endShape();        
}
